### ISSUE TITLE
Ambient L4 ztunnel Stability Solution

### ISSUE THREAD

**Issue by zengyuxing007** [istio/istio#48949]
**Describe the feature request**

In ambient mode, ztunnel is deployed using daemonset, which is a single point for a single node. If ztunnel crashes, the traffic of the whole node is affected.   Is there a fallback option or a more stable solution?

The ztunnel pod does not have a lower [eviction priority](https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/).   this means that when the node node is low on resources, it is possible that the ztunnel is evicted, but the business pod is still alive and traffic delivery fails.

In addition, when upgrading the version of ztunnel, we need to consider the impact of the upgrade process on the business traffic, whether there is a zero downtime solution ？

**Describe alternatives you've considered**

- [X] Ambient
 - [X] Docs
  [ ] Dual Stack
- [X] Installation
-  [X] Networking
 - [X] Performance and Scalability
  [ ] Extensions and Telemetry
  [ ] Security
  [ ] Test and Release
- [X] User Experience

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

**Comment by linsun** [comment]
Thanks for opening this! This could be considered after the initial ambient beta if folks are interested in fixing it.

**Comment by zengyuxing007** [comment]
We've had some discussion in this issue https://github.com/istio/istio/issues/48387, and it's a known issue that there are some traffic disruptions with the current ztunnel upgrade(in-place).

**Comment by linsun** [comment]
I saw you added ambient beta label, defer to @istio/technical-oversight-committee to weigh in, + @keithmattix 

My intuition is this is a not a beta blocker. Users should have plan for HA for their workloads and not have everything deployed in 1 node in case ztunnel on that node crashes.

**Comment by nrjpoddar** [comment]
Shouldn't be a beta blocker as if we alternate ways to achieve HA and we get this added after user feedback which we can get with promotion to beta.

**Comment by linsun** [comment]
Thanks Neeraj, i'll wait for others to chime in before removing the beta label.

**Comment by zengyuxing007** [comment]
I have tried to solve the problem of request failures (5xx) when ztunnel is upgraded or restarted, and because of the current ztunnel HBONE protocol implementation, this problem is basically unsolvable.

The main reason is that the current HBONE implementation is based on a tcp connection (tcp over http2+tls), and when the ztunnel process exits, the tcp connection or tunnel is broken and cannot be recovered.

Do we consider something like cilium's mTLS + socket LB solution as an optional L4 implementation?  @howardjohn @linsun @bleggett 

The user can decide to choose a different L4 implementation by installing a different configuration (ztunnel or ztunnel-less (eBPF) ). 

I am glad to submit a corresponding proposal and related implementation if possible.

**Comment by linsun** [comment]
@zengyuxing007 what John said. May I ask why you'd be interested in an optional L4 impl? Is "request failures (5xx) when ztunnel is upgraded or restarted" the main reason?

**Comment by zengyuxing007** [comment]
> @zengyuxing007 what John said. May I ask why you'd be interested in an optional L4 impl? Is "request failures (5xx) when ztunnel is upgraded or restarted" the main reason?
@linsun  Yes, we expect Ambient to be available for production environments, and Your know, Mesh as infrastructure should not cause business downtime when upgraded. Although this downtime is relatively short, it will always cause some trouble for the business side and make the end user experience not so good.

**Comment by zengyuxing007** [comment]
> Cilium does not have an mTLS solution. See https://istio.io/latest/blog/2023/secure-apps-with-istio/.
> 
> Its plausible we could offer an insecure mode where we only have waypoints over plaintext I suppose, but its not really a top priority given the security priority in the project.

@howardjohn  istio's mTLS implementation is recognized as more secure. [cilium's mTLS](https://isovalent.com/blog/post/2022-05-03-servicemesh-security/) is just kind of what they call it, but whatever it can also be a zero-trust implementation, which does have some argument.

Solo has a blog talk about this: https://www.solo.io/blog/could-network-cache-based-identity-be-mistaken/ .
There's also some discussion on Slack. (https://cilium.slack.com/archives/C02QKQDTVDX/p1701466174930209)


> Its plausible we could offer an insecure mode where we only have waypoints over plaintext I suppose, but its not really a top priority given the security priority in the project.

Yes,  The current HBONE protocol is supposed to require TLS , although PeerAuthentication is configured as DISABLE. 

mainly because the current HBONE tunnel is not repairable or stateless.

If the HBONE protocol can be designed to be more loose and can based on UDP, e.g. UDP Tunnel, this problem can be solved, but of course it depends on the envoy side to be able to adapt the modification accordingly. I'll try to see if there is a relevant way to do this, thanks~

**Comment by zengyuxing007** [comment]
Waypoint Sandwich mode seems like it would make this easier.

Related Issue: https://github.com/istio/istio/issues/48362

**Comment by zengyuxing007** [comment]
We have implemented a solution for the high availability issue of ztunnel in the following way.

POC test that is ok~  When ztunnel  upgraded or restarted, the traffic is lossless.

![image](https://github.com/istio/istio/assets/370590/d922809b-90e8-4800-b9e1-9ee2006a438b)

The implement involves some modifications to ztunnel and istio-cni, and I am wondering if the ambient has considered supporting different tunnel protocols. If so, I am glad to  submit a more detailed proposal.

In the long run, this may also involve the scalability of ztunnel,ref issue:  https://github.com/istio/ztunnel/issues/852


cc @howardjohn @bleggett @costinm @linsun

**Comment by keithmattix** [comment]
Might be worth looking into masque for this

**Comment by Nova-Logic** [comment]
it should not be staled
